### PR TITLE
make: docker: Run docker test in parallel for all jobs

### PR DESCRIPTION
**Pull request by jcvenegas**
Running test in serial takes a long time, just qemu uses today
parallel, refactor makefile to run all as parallel.

Fixes: #2426

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -91,36 +91,42 @@ endif
...
-		./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
-else ifneq (${FOCUS},)
-	./ginkgo -failFast -v -focus "${FOCUS}" -skip "${SKIP}" \
-		./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
-else ifeq (centos7,$(ID)$(VERSION_ID))
-# Run tests sequentially, parallel tests fail randomly in Centos 7
-	./ginkgo -failFast -v -skip "${SKIP}" \
-		./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT}
+ifeq (false,$(RUN_PARALLEL))
+	@echo "Running test in serial, this will take serverala minutes"
```
Comment (jodh-intel) [inline Makefile:122]: Typo: "several"

File: Makefile [hunk 2]
```diff
@@ -91,36 +91,42 @@ endif
...
+
+ifeq ($(KATA_HYPERVISOR),firecracker)
+GINKGO_FLAGS=-skip "${SKIP_FIRECRACKER}"
+endif
+
+ifeq ($(KATA_HYPERVISOR),cloud-hypervisor)
+GINKGO_FLAGS=-skip "${SKIP_CLH}"
+endif
+
+RUN_PARALLEL :=
```
Comment (jodh-intel) [inline Makefile:107]: I can see that the code will work, but I'd still rather we default this to `true` as currently we're saying "if the variable is unset, it's true which is slightly counter-intuitive.

**Comment by grahamwhaley** [comment]
looks to be about 50/50 on CI success rate....
Didn't we try parallel running tests in the CI in the past, and back it out due to spurious race/lock/tmpfile clashes? /cc @chavafg @devimc

**Comment by devimc** [comment]
@grahamwhaley 

> Didn't we try parallel running tests in the CI in the past, and back it out due to spurious race/lock/tmpfile clashes?

yes, the tests that can't run in parallel are marked as "[serial test]"

**Comment by jcvenegas** [comment]
@devimc @grahamwhaley @jodh-intel I fixed a few things, I needed to use an extra Makefiles magic let me know if you have some feedback.

File: Makefile [hunk 3]
```diff
@@ -91,36 +91,68 @@ endif
...
+GINKGO_FLAGS := -failFast -v -focus "${FOCUS}" $(EXTRA_GINKGO_FLAGS)
+
+GINKGO_TEST_FLAGS := ./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT} -hypervisor=$(KATA_HYPERVISOR)
+
+# Append strings to skip for ginkgo using '|'
+define append_skip_ginkgo
+$(info adding $(2) to $(1))
+$(if $(SKIP_GINKGO),$(eval $(1)= $(SKIP_GINKGO)|))
+$(if $(2),$(eval $(1) = $($(1))$(2)))
+$(info Oki=$($(1)))
```
Comment (jodh-intel) [inline Makefile:105]: Is this line correct?

File: Makefile [hunk 4]
```diff
@@ -91,36 +91,67 @@ endif
...
 
+EXTRA_GINKGO_FLAGS :=
+SKIP_GINKGO =
+GINKGO_FLAGS := -failFast -v -focus "${FOCUS}" $(EXTRA_GINKGO_FLAGS)
+
+GINKGO_TEST_FLAGS := ./integration/docker/ -- -runtime=${RUNTIME} -timeout=${TIMEOUT} -hypervisor=$(KATA_HYPERVISOR)
+
+# Append strings to skip for ginkgo using '|'
+define append_skip_ginkgo
+$(info adding $(2) to $(1))
```
Comment (jodh-intel) [inline Makefile:102]: Idea: You could add an "eye-catcher" tag this:

```make
$(info "INFO: Adding $(2) to $(1))
```

**Comment by jcvenegas** [comment]
Some numbers running in pareallel, I calculated  times for cloud-hypervisor CI, where we go. from 35  to 17 minutes (just  docker testing, setup is not consider here). Something I dont get is why there is one more skipped test in the case of firecracker I imagine that is a conflict the the new regex to skip of focus in serial cases. I consider I will keep fc to run in serial for now.

Firecracker:
Before:
http://jenkins.katacontainers.io/job/kata-containers-tests-ubuntu-1804-PR-firecracker/1316/consoleFull

```
SUCCESS! -- 57 Passed | 0 Failed | 0 Pending | 193 Skipped
```
After:
http://jenkins.katacontainers.io/job/kata-containers-tests-ubuntu-1804-PR-firecracker/1319/consoleFull
```
SUCCESS! -- 17 Passed | 0 Failed | 0 Pending | 66 Skipped
SUCCESS! -- 14 Passed | 0 Failed | 0 Pending | 22 Skipped
SUCCESS! -- 11 Passed | 0 Failed | 0 Pending | 37 Skipped
SUCCESS! -- 14 Passed | 0 Failed | 0 Pending | 69 Skipped
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 250 Skipped
Total: 56
Skipped: 94
```

Cloud-hypervisor:
Before:
http://jenkins.katacontainers.io/job/kata-containers-tests-ubuntu-1804-PR-cloud-hypeprvisor-docker/30/consoleFull
```
SUCCESS! -- 233 Passed | 0 Failed | 0 Pending | 17 Skipped
00:05:16.371 : Start time
00:40:46.359: Stop Time
Final time: 35:30
```
After:
http://jenkins.katacontainers.io/job/kata-containers-tests-ubuntu-1804-PR-cloud-hypeprvisor-docker/31/console
```
SUCCESS! -- 58 Passed | 0 Failed | 0 Pending | 4 Skipped
SUCCESS! -- 52 Passed | 0 Failed | 0 Pending | 1 Skipped
SUCCESS! -- 53 Passed | 0 Failed | 0 Pending | 6 Skipped
SUCCESS! -- 57 Passed | 0 Failed | 0 Pending | 19 Skipped
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 237 Skipped
00:10:37.534: Start time
00:27:50.774: Stop time
Final time: 17:13
Total: 233
```

**Comment by jcvenegas** [comment]
Restating docker job seems that a random issue, lets save some logs here:
```
00:22:20.762 [1] [Fail] capabilities drop and add capabilities [It] cap_mknod 
00:22:20.762 [1] /tmp/jenkins/workspace/kata-containers-tests-ubuntu-1804-PR-cloud-hypeprvisor-docker/go/src/github.com/kata-containers/tests/integration/docker/cap_test.go:47
00:22:20.762 [1] 
00:22:20.762 [1] Ran 35 of 123 Specs in 679.424 seconds
00:22:20.762 [1] FAIL! -- 34 Passed | 1 Failed | 0 Pending | 88 Skipped
```

**Pull request merged**